### PR TITLE
DCOS-12802: Introduce Jenkinsfile

### PR DESCRIPTION
**Pull request by orlandohohmeier**
Introduce Jenkinsfile to use Jenkins Pipelines and track and test different build/test configurations.
The current configuration will provision an agent using the `Dockerfile` and then lint, build and test the UI.

Closes DCOS-12802

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    stage('Build') {
+      steps {
+        ansiColor('xterm') {
+          sh '''npm run build-assets'''
+        }
+        stash includes: 'dist/*', name: 'dist'
+      }
+    }
+
+    stage('Integration Tests') {
```
Comment (weblancaster) [inline Jenkinsfile]: I believe we would benefit a lot in performance if we use the `parallel` feat from jenkins pipeline.
The places I believe we can use is `Lint` `Unit test` `Integration tests` and `System test`

What you think @orlandohohmeier 

for more info: https://www.cloudbees.com/blog/parallelism-and-distributed-builds-jenkins

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    dockerfile true
+  }
+
+  environment {
+    JENKINS_VERSION = 'yes'
+    NODE_PATH = 'node_modules'
+  }
+
+  options {
+    timeout(time: 2, unit: 'HOURS')
```
Comment (weblancaster) [inline Jenkinsfile]: 2hrs is incredibly long, especially for our code base. I would expect 1hr max.

Comment (weblancaster) [reply Jenkinsfile]: If there is a specific stage that we need to timeout we could only add to that stage.

**Comment by weblancaster** [comment]
@orlandohohmeier I left some comments, other than that looks good and I'm excited for us to introduce Jenkinsfile pipeline (especially if we decrease the time it takes to run).

File: Jenkinsfile (hunk 1 above)
Comment (orlandohohmeier) [reply Jenkinsfile]: @weblancaster I had a version that used parallel steps, but ran into memory issues as all of these test are executed on the same machine if we don't use the `node` directive to schedule the work on a different nodes and I haven't figured out how to share the initially build docker image which results in rather long bootstrap times and thus is not faster than running these synchronously. Anyway, I'll continue investigating ways to improve the performance and build time and really appreciate your comments and ideas.

File: Jenkinsfile (hunk 2 above)
Comment (orlandohohmeier) [reply Jenkinsfile]: You're right that 2 hours is quite long and our builds should complete after 30 mins on average. I'll reduce the general timeout to 1 hour as it should be ok to flag a build that takes longer as a failure.

File: Jenkinsfile (hunk 1 above)
Comment (weblancaster) [reply Jenkinsfile]: Ok cool, thanks for the feedback.

File: Jenkinsfile (hunk 2 above)
Comment (weblancaster) [reply Jenkinsfile]: Thanks!

**Comment by nLight** [comment]
What do we need to do to make this happen?

**Comment by orlandohohmeier** [comment]
@nLight I need to further investigate why our jest workers (unit tests) or the chrome renderer (integration tests) fail. I thought it's a resource issue, but the agent should have enough resources. I'll continue to experiment with different configurations and will look into ways to better debug the failure. If anyone has an idea how to solve the issues please don't hesitate to adjust our Jenkins setup or this PR.

File: Dockerfile [hunk 3]
```diff
@@ -33,6 +33,10 @@ RUN set -x \
   # Post-install java certificates
   /var/lib/dpkg/info/ca-certificates-java.postinst configure \
 
+ # Install npm dependencies
+  && cd /dcos-ui \
+  && npm install -g git://github.com/johntron/http-server.git#proxy-secure-flag \
```
Comment (nLight) [inline Dockerfile]: I'd rather see python3 http-server here...

**Comment by nLight** [comment]
`ENOMEM: not enough memory, read` for linting command

**Comment by nLight** [comment]
Alfred test this

**Comment by orlandohohmeier** [comment]
Alfred could you please retest this PR?

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      post {
+        always {
+          archiveArtifacts 'cypress/**/*'
+          junit 'cypress/*.xml'
+        }
+      }
+    }
+
+
+//    stage('System Test') {
```
Comment (nLight) [inline Jenkinsfile]: It's ok to have this part commented out

**Comment by nLight** [comment]
@weblancaster please consider ✅

**Pull request merged**